components/server/ee/src/billing/entitlement-service.ts

```diff
@@ -4,52 +4,103 @@
  * See License-AGPL.txt in the project root for license information.
  */
 
-import { User, WorkspaceInstance, WorkspaceTimeoutDuration } from "@gitpod/gitpod-protocol";
+import {
+    User,
+    WorkspaceInstance,
+    WorkspaceTimeoutDuration,
+    WORKSPACE_TIMEOUT_DEFAULT_LONG,
+} from "@gitpod/gitpod-protocol";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
 import { inject, injectable } from "inversify";
-import { EntitlementService } from "../../../src/billing/entitlement-service";
+import { EntitlementService, MayStartWorkspaceResult } from "../../../src/billing/entitlement-service";
 import { Config } from "../../../src/config";
-import { MayStartWorkspaceResult } from "../user/eligibility-service";
+import { BillingModes } from "./billing-mode";
 import { EntitlementServiceChargebee } from "./entitlement-service-chargebee";
 import { EntitlementServiceLicense } from "./entitlement-service-license";
+import { EntitlementServiceUBP } from "./entitlement-service-ubp";
 
 /**
  * The default implementation for the Enterprise Edition (EE). It decides based on config which ruleset to choose for each call.
+ *
+ * As a last safety net for rolling this out, it swallows all errors and turns them into log statements.
  */
 @injectable()
 export class EntitlementServiceImpl implements EntitlementService {
     @inject(Config) protected readonly config: Config;
-    @inject(EntitlementServiceChargebee) protected readonly etsChargebee: EntitlementServiceChargebee;
-    @inject(EntitlementServiceLicense) protected readonly etsLicense: EntitlementServiceLicense;
+    @inject(BillingModes) protected readonly billingModes: BillingModes;
+    @inject(EntitlementServiceChargebee) protected readonly chargebee: EntitlementServiceChargebee;
+    @inject(EntitlementServiceLicense) protected readonly license: EntitlementServiceLicense;
+    @inject(EntitlementServiceUBP) protected readonly ubp: EntitlementServiceUBP;
 
     async mayStartWorkspace(
         user: User,
-        date: Date,
+        date: Date = new Date(),
         runningInstances: Promise<WorkspaceInstance[]>,
     ): Promise<MayStartWorkspaceResult> {
-        if (!this.config.enablePayment) {
-            return await this.etsLicense.mayStartWorkspace(user, date, runningInstances);
+        try {
+            const billingMode = await this.billingModes.getBillingModeForUser(user, date);
+            switch (billingMode.mode) {
+                case "none":
+                    return this.license.mayStartWorkspace(user, date, runningInstances);
+                case "chargebee":
+                    return this.chargebee.mayStartWorkspace(user, date, runningInstances);
+                case "usage-based":
+                    return this.ubp.mayStartWorkspace(user, date, runningInstances);
+            }
+        } catch (err) {
+            log.error({ userId: user.id }, "EntitlementService error: mayStartWorkspace", err);
+            return {};
```

Comment:
@jankeromnes, wdyt, if anything goes wrong (e.g. in UB computation) this basically means, no limits are enforced. the question is, do we want it to fail silently in error cases?

Reply:
I agree that it seems better to reject `mayStartWorkspace` if there is an error here, instead of failing silently.

---

Also, the `MayStartWorkspaceResult` type seems quite weird to me:

- Apparently it's an object that contains two keys, `enoughCredits: boolean` and `hasHitParallelWorkspaceLimit`? Are no other fields needed?

https://github.com/gitpod-io/gitpod/blob/678a6a2f0cd1a82471b5a6a75946adcce2eafcc6/components/server/ee/src/user/eligibility-service.ts#L18-L21

- Also, reading the code, it seems that we occasionally deviate from this type, e.g. by returning an empty object (`{}` here), or `undefined` (if `billingMode.mode` becomes `undefined` for some reason one day)

Reply:
thanks for the overview @jankeromnes!

revising this shortly. 

also, the base (empty) implementation in non-ee folder is clear, which supports your comment.

Reply:
@jankeromnes can you have a look at https://github.com/gitpod-io/gitpod/pull/11936/commits/bcb06f44877897302e5ec6b89d66a17d771dee48

I hope to have addressed the bullet points.